src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs

```diff
@@ -83,32 +83,79 @@ internal partial class CodeFixService : ForegroundThreadAffinitizedObject, ICode
 
             using (var diagnostics = SharedPools.Default<List<DiagnosticData>>().GetPooledObject())
             {
-                var fullResult = await _diagnosticService.TryAppendDiagnosticsForSpanAsync(document, range, diagnostics.Object, cancellationToken: cancellationToken).ConfigureAwait(false);
-                foreach (var diagnostic in diagnostics.Object)
+                using (var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken))
                 {
-                    cancellationToken.ThrowIfCancellationRequested();
+                    var linkedToken = linkedTokenSource.Token;
 
-                    if (!range.IntersectsWith(diagnostic.TextSpan))
+                    // This flag is used by SuggestedActionsSource to track what solution is was
+                    // last able to get "full results" for.
+                    var isFullResult = await _diagnosticService.TryAppendDiagnosticsForSpanAsync(
+                        document, range, diagnostics.Object, cancellationToken: linkedToken).ConfigureAwait(false);
+
+                    var errorDiagnostics = diagnostics.Object.Where(d => d.Severity == DiagnosticSeverity.Error);
+                    var otherDiagnostics = diagnostics.Object.Where(d => d.Severity != DiagnosticSeverity.Error);
+
+                    // Kick off a task that will determine there's an Error Diagnostic with a fixer
+                    var errorDiagnosticsTask = Task.Run(
+                        () => GetFirstDiagnosticWithFixAsync(document, errorDiagnostics, range, linkedToken),
+                        linkedToken);
+
+                    // Kick off a task that will determine if any non-Error Diagnostic has a fixer
+                    var otherDiagnosticsTask = Task.Run(
+                        () => GetFirstDiagnosticWithFixAsync(document, otherDiagnostics, range, linkedToken),
+                        linkedToken);
+
+                    var firstCompletedTask = await Task.WhenAny(errorDiagnosticsTask, otherDiagnosticsTask).ConfigureAwait(false);
+                    if (firstCompletedTask == errorDiagnosticsTask)
                     {
-                        continue;
-                    }
+                        if (firstCompletedTask.Result != null)
+                        {
+                            // Error task completed first and had a result; cancel pending work and return
+                            linkedTokenSource.Cancel();
+                            return CreateResult(isFullResult, firstCompletedTask.Result);
```

Comment:
calling .Result is unsafe.  firstCompletedTask may have completed for any reason (including cancellation).  You need to do 'await firstCompletedTask...' here.  (this also applies to many cases below).